ops_tests/test_upgrade_charm.py

```diff
@@ -0,0 +1,78 @@
+import logging
+import subprocess
+import tempfile
+import asyncio
+
+from juju.errors import JujuError
+import yaml
+
+import pytest
+
+log = logging.getLogger(__name__)
+
+
+@pytest.mark.abort_on_fail
+async def test_deploy_old_bundle(ops_test):
+    await ops_test.model.deploy("cs:kubeflow")
+    subprocess.run(
+        [
+            'juju',
+            'kubectl',
+            'patch',
+            'role/istio-ingressgateway-operator',
+            '-p',
+            yaml.dump(
+                {
+                    "apiVersion": "rbac.authorization.k8s.io/v1",
+                    "kind": "Role",
+                    "metadata": {"name": "istio-ingressgateway-operator"},
+                    "rules": [{"apiGroups": ["*"], "resources": ["*"], "verbs": ["*"]}],
+                }
+            ),
+        ]
+    )
+    # raise_on_error=False due to kubeflow-profiles going to an error state if
+    # kubeflow-dashboard CRDs aren't created yet.
+    await ops_test.model.wait_for_idle(raise_on_error=False)
+
+
+async def test_upgrade_apps_in_bundle(ops_test):
+    res = subprocess.run(["charm", "-v"])
```

Comment:
Can we include `charm` in the list of pip packages to install? That way tox will install it automatically

Reply:
pytest-operator expects the charm snap to be installed and will error out if it's not. I tested it with only the pip package and seems to specifically expect the snap. So my check is probably unnecessary anyway.